server/models/service-process.js

```diff
@@ -169,18 +170,37 @@ module.exports = function extendServiceProcess(ServiceProcess) {
   }
   ServiceProcess.recordStatusWdUpdate = recordStatusWdUpdate;
 
-  function _findProcess(instanceId, workerId, pid, pst) {
-    var filter = {
-      where: {
-        serviceInstanceId: instanceId,
-        workerId: +workerId,
-        pid: pid,
-      }
+  function _findOrCreateProcess(instanceId, workerId, pid, pst, callback) {
+    var procData = {
+      pid: pid,
+      workerId: workerId,
+      serviceInstanceId: instanceId,
+      startTime: new Date(pst),
     };
-    if (pst) {
-      filter.where.startTime = new Date(pst);
-    }
-    return ServiceProcess.findOne.bind(ServiceProcess, filter);
+    var filter = {where: procData};
+    debug('_findOrCreateProcess: %j', procData);
+
+    // Try to find an existing process or create a new one. If an error occurs,
+    // it might be a duplicate key error, in which case, the process was already
+    // created by a parallel invocation, attempt to find that process.
+
+    // XXX(KR) The processes create does not include ppid, which may cause race
+    // conditions during cleanup due to supervisor exit.
```

Comment:
is there anything we can do to fix this? should we make sure we cleanup on start any processes which don't have a ppid, since they will never get one? Or, is that only true for the pm use-case, not central?